Add snow DA update and recentering for the EnKF forecasts

CoryMartin-NOAA opened the pull request:
# Description
Apologies in that this is a fairly substantial PR, but a lot of things were needed to make this happen. This PR adds the capability to update the ensemble of snow states by recentering the ensemble mean to the deterministic snow analysis and applying increments as appropriate.

  Resolves #2585 

# Type of change
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? YES in the sense that results will change
- Does this change require a documentation update? Maybe?

# How has this been tested?
- Cycled test on Hera

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

CoryMartin-NOAA committed:
start to add new task

CoryMartin-NOAA committed:
add new j-job

CoryMartin-NOAA committed:
Save to do some initial testing

CoryMartin-NOAA committed:
more updates

CoryMartin-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
end of day commit

CoryMartin-NOAA committed:
End of day/week commit

CoryMartin-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
save

CoryMartin-NOAA committed:
Merge branch 'feature/snow-ensemble' of https://github.com/CoryMartin-NOAA/global-workflow into feature/snow-ensemble

CoryMartin-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
snow changes

CoryMartin-NOAA committed:
end of day commit

CoryMartin-NOAA committed:
stage files is working

CoryMartin-NOAA committed:
save updates

CoryMartin-NOAA committed:
more updates

CoryMartin-NOAA committed:
make norm happy

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
norm is an angry old man

CoryMartin-NOAA committed:
remove esnowanl from enkfgfs

CoryMartin-NOAA committed:
Merge branch 'feature/snow-ensemble' of https://github.com/CoryMartin-NOAA/global-workflow into feature/snow-ensemble

CoryMartin-NOAA committed:
end of day commit

CoryMartin-NOAA committed:
fregrid working but seems a bit hacky

CoryMartin-NOAA committed:
norms and remove commented out code

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
update snow ens

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
update some submodules

CoryMartin-NOAA committed:
update some things

CoryMartin-NOAA committed:
Merge branch 'feature/snow-ensemble' of https://github.com/CoryMartin-NOAA/global-workflow into feature/snow-ensemble

CoryMartin-NOAA committed:
update snow stage yaml

CoryMartin-NOAA committed:
whoops

CoryMartin-NOAA committed:
hack increment times

CoryMartin-NOAA committed:
dates for increment

CoryMartin-NOAA committed:
fregrid names

CoryMartin-NOAA committed:
updates to test

CoryMartin-NOAA committed:
fix method call

CoryMartin-NOAA committed:
Merge branch 'feature/snow-ensemble' of https://github.com/CoryMartin-NOAA/global-workflow into feature/snow-ensemble

CoryMartin-NOAA committed:
coding norms

CoryMartin-NOAA committed:
end of week commit

CoryMartin-NOAA committed:
one more end of week commit

CoryMartin-NOAA committed:
smore changes

CoryMartin-NOAA committed:
Merge branch 'feature/snow-ensemble' of https://github.com/CoryMartin-NOAA/global-workflow into feature/snow-ensemble

CoryMartin-NOAA committed:
ensmean to ensstat

CoryMartin-NOAA committed:
IAU changes

CoryMartin-NOAA committed:
coding norms + more IAU fixes

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
Update gdas

CoryMartin-NOAA committed:
update enkf sfc update

CoryMartin-NOAA committed:
more changes

CoryMartin-NOAA committed:
update ci case

CoryMartin-NOAA committed:
fix dependencies

CoryMartin-NOAA committed:
ignore these bufr2ioda links

CoryMartin-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
fixing some bugs

CoryMartin-NOAA committed:
Merge branch 'feature/snow-ensemble' of https://github.com/corymartin-noaa/global-workflow into feature/snow-ensemble

CoryMartin-NOAA commented:
@jiaruidong2017 I've tested this but can you try it out and see if you find any issues running under a different account from my own. Thanks!

jiaruidong2017 commented:
> @jiaruidong2017 I've tested this but can you try it out and see if you find any issues running under a different account from my own. Thanks!

@CoryMartin-NOAA I will try it, and let you know then.

CoryMartin-NOAA committed:
shellcheck

CoryMartin-NOAA commented:
Reviewers, please note that this PR currently points to a hash of `gdas.cd` which is not in develop yet, so this should not be merged until the `gdas.cd` hash is updated.

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
more shellcheck

jiaruidong2017 commented:
@CoryMartin-NOAA Would you please tell me where is your experiment directory for this PR? Thanks.

CoryMartin-NOAA commented:
> @CoryMartin-NOAA Would you please tell me where is your experiment directory for this PR? Thanks.

@jiaruidong2017 on Hera see `/scratch1/NCEPDEV/stmp2/Cory.R.Martin/tmp_june2024/snowens/exp/snowenstest`

jiaruidong2017 commented:
> @jiaruidong2017 I've tested this but can you try it out and see if you find any issues running under a different account from my own. Thanks!

@CoryMartin-NOAA I have cloned your repo and built the global-workflow under your branch of `CoryMartin-NOAA:feature/snow-ensemble`. Then, I setup the experiment, and conducted the test run. The test run was technically successful. Therefore, this PR is ready to be reviewed.

DavidHuber-NOAA commented on ci/cases/pr/C96_atmaerosnowDA.yaml:
The name of this test should be updated to indicate it is now hybvar.  I suggest `ci/cases/pr/C96C48_hybatmaerosnowDA.yaml`.

DavidHuber-NOAA commented on env/HERA.env:
```diff
@@ -119,6 +119,16 @@ elif [[ "${step}" = "snowanl" ]]; then
+    [[ ${NTHREADS_ESNOWANL} -gt ${nth_max} ]] && export NTHREADS_ESNOWANL=${nth_max}
+    export APRUN_ESNOWANL="${launcher} -n ${npe_esnowanl} --cpus-per-task=${NTHREADS_ESNOWANL}"
+
+    export APRUN_APPLY_INCR="${launcher} -n 6"
```
Is `-n 6` fixed for all resolutions for `APPLY_INCR`?  In any case, I suggest declaring `npe_apply_incr` in `config.resources` and referencing it in each `env` file.  This should probably also be set for the `snowanl` job.

CoryMartin-NOAA replied:
Done

DavidHuber-NOAA commented on parm/gdas/snow_stage_ens_update.yaml.j2:
```diff
@@ -0,0 +1,113 @@
+{% set bkg_time = SNOW_WINDOW_BEGIN | to_fv3time %}
+{% else %}
+{% set bkg_time = current_cycle | to_fv3time %}
+{% endif %}
```
Suggest tabbing for readability:
```suggestion
{% if DOIAU %}
    {% set bkg_time = SNOW_WINDOW_BEGIN | to_fv3time %}
{% else %}
    {% set bkg_time = current_cycle | to_fv3time %}
{% endif %}
```

DavidHuber-NOAA commented on parm/gdas/snow_finalize_ens_update.yaml.j2:
```diff
@@ -0,0 +1,70 @@
+{% for tile in range(1, 7) %}
+- ["{{ DATA }}/inc/ensmean/snowinc.{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ com_ns.COM_SNOW_ANALYSIS_MEM }}/snowinc.{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
+{% endfor %}
+{% endif %}
```
Suggest tabbing for readability:
```suggestion
{% for mem in range(1, NMEM_ENS + 1) %}
    # define variables
    # Declare a dict of search and replace terms to run on each template
    {% set tmpl_dict = {'ROTDIR':ROTDIR,
                        'RUN':RUN,
                        'YMD':PDY,
                        'HH':cyc,
                        'MEMDIR':"mem" + '%03d' % mem} %}

    # Replace template variables with tmpl_dict, one key at a time
    # This must be done in a namespace to overcome jinja scoping
    #  Variables set inside of a for loop are lost at the end of the loop
    #  unless they are part of a namespace
    {% set com_ns = namespace(COM_SNOW_ANALYSIS_MEM = COM_SNOW_ANALYSIS_TMPL) %}

    {% for key in tmpl_dict.keys() %}
        {% set search_term = '${' + key + '}' %}
        {% set replace_term = tmpl_dict[key] %}
        {% set com_ns.COM_SNOW_ANALYSIS_MEM = com_ns.COM_SNOW_ANALYSIS_MEM.replace(search_term, replace_term) %}
    {% endfor %}
    {% for tile in range(1, 7) %}
- ["{{ DATA }}/anl/mem{{ '%03d' % mem }}/{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ com_ns.COM_SNOW_ANALYSIS_MEM }}/{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
    {% endfor %}
    {% if DOIAU %}
        # if using IAU, also need analyses copied at the beginning of the window
        {% for tile in range(1, 7) %}
- ["{{ DATA }}/anl/mem{{ '%03d' % mem }}/{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ com_ns.COM_SNOW_ANALYSIS_MEM }}/{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
        {% endfor %}
    {% endif %}
{% endfor %}
######################################
# copy ensemble mean increment to COM
######################################
# define variables
# Declare a dict of search and replace terms to run on each template
{% set tmpl_dict = {'ROTDIR':ROTDIR,
                    'RUN':RUN,
                    'YMD':PDY,
                    'HH':cyc,
                    'MEMDIR':"ensstat"} %}

# Replace template variables with tmpl_dict, one key at a time
# This must be done in a namespace to overcome jinja scoping
#  Variables set inside of a for loop are lost at the end of the loop
#  unless they are part of a namespace
{% set com_ns = namespace(COM_SNOW_ANALYSIS_MEM = COM_SNOW_ANALYSIS_TMPL) %}

{% for key in tmpl_dict.keys() %}
    {% set search_term = '${' + key + '}' %}
    {% set replace_term = tmpl_dict[key] %}
    {% set com_ns.COM_SNOW_ANALYSIS_MEM = com_ns.COM_SNOW_ANALYSIS_MEM.replace(search_term, replace_term) %}
{% endfor %}
{% for tile in range(1, 7) %}
- ["{{ DATA }}/inc/ensmean/snowinc.{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ com_ns.COM_SNOW_ANALYSIS_MEM }}/snowinc.{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
{% endfor %}
{% if DOIAU %}
    # if using IAU, also need increment copied at the beginning of the window
    {% for tile in range(1, 7) %}
- ["{{ DATA }}/inc/ensmean/snowinc.{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ com_ns.COM_SNOW_ANALYSIS_MEM }}/snowinc.{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
    {% endfor %}
{% endif %}
```

DavidHuber-NOAA commented on parm/gdas/snow_stage_ens_update.yaml.j2:
```diff
@@ -0,0 +1,113 @@
+{% set search_term = '${' + key + '}' %}
+{% set replace_term = tmpl_dict[key] %}
+{% set com_ns.COM_SNOW_ANALYSIS_MEM = com_ns.COM_SNOW_ANALYSIS_MEM.replace(search_term, replace_term) %}
+{% endfor %}
```
```suggestion
{% for key in tmpl_dict.keys() %}
    {% set search_term = '${' + key + '}' %}
    {% set replace_term = tmpl_dict[key] %}
    {% set com_ns.COM_SNOW_ANALYSIS_MEM = com_ns.COM_SNOW_ANALYSIS_MEM.replace(search_term, replace_term) %}
{% endfor %}
```

DavidHuber-NOAA commented on parm/gdas/snow_stage_ens_update.yaml.j2:
```diff
@@ -0,0 +1,113 @@
+- ["{{ com_prev_ns.COM_ATMOS_RESTART_MEM }}/{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ DATA }}/anl/mem{{ '%03d' % mem }}/{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
+{% endfor %}
+{% endif %}
+{% endfor %}
```
```suggestion
{% for mem in range(1, NMEM_ENS + 1) %}
    # define variables
    # Declare a dict of search and replace terms to run on each template
    {% set tmpl_dict = {'ROTDIR':ROTDIR,
                        'RUN':RUN,
                        'YMD':gPDY,
                        'HH':gcyc,
                        'MEMDIR':"mem" + '%03d' % mem} %}

    # Replace template variables with tmpl_dict, one key at a time
    # This must be done in a namespace to overcome jinja scoping
    #  Variables set inside of a for loop are lost at the end of the loop
    #  unless they are part of a namespace
    {% set com_prev_ns = namespace(COM_ATMOS_RESTART_MEM = COM_ATMOS_RESTART_TMPL) %}

    {% for key in tmpl_dict.keys() %}
        {% set search_term = '${' + key + '}' %}
        {% set replace_term = tmpl_dict[key] %}
        {% set com_prev_ns.COM_ATMOS_RESTART_MEM = com_prev_ns.COM_ATMOS_RESTART_MEM.replace(search_term, replace_term) %}
    {% endfor %}
    # we need to copy them to two places, one serves as the basis for the analysis
    {% for tile in range(1, 7) %}
- ["{{ com_prev_ns.COM_ATMOS_RESTART_MEM }}/{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ DATA }}/bkg/mem{{ '%03d' % mem }}/{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
- ["{{ com_prev_ns.COM_ATMOS_RESTART_MEM }}/{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ DATA }}/anl/mem{{ '%03d' % mem }}/{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
    {% endfor %}
    {% if DOIAU %}
        # if using IAU, also need backgrounds copied at the beginning of the window
        # we need to copy them to two places, one serves as the basis for the analysis
        {% for tile in range(1, 7) %}
- ["{{ com_prev_ns.COM_ATMOS_RESTART_MEM }}/{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ DATA }}/bkg/mem{{ '%03d' % mem }}/{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
- ["{{ com_prev_ns.COM_ATMOS_RESTART_MEM }}/{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ DATA }}/anl/mem{{ '%03d' % mem }}/{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
        {% endfor %}
    {% endif %}
{% endfor %}
```

DavidHuber-NOAA commented on parm/gdas/snow_finalize_ens_update.yaml.j2:
```diff
@@ -0,0 +1,70 @@
+{% for tile in range(1, 7) %}
+- ["{{ DATA }}/anl/mem{{ '%03d' % mem }}/{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ com_ns.COM_SNOW_ANALYSIS_MEM }}/{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
+{% endfor %}
+{% if DOIAU %}
```
Suggest checking for `True`.  If `DOIAU` is undefined, it will be set as `Undefined`, which will evaluate as `True`.
```suggestion
{% if DOIAU == True %}
```

CoryMartin-NOAA replied:
Thanks for this, I didn't know this about Jinja

DavidHuber-NOAA commented on parm/gdas/snow_finalize_ens_update.yaml.j2:
```diff
@@ -0,0 +1,70 @@
+{% if DOIAU %}
+# if using IAU, also need increment copied at the beginning of the window
+{% for tile in range(1, 7) %}
+- ["{{ DATA }}/inc/ensmean/snowinc.{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ com_ns.COM_SNOW_ANALYSIS_MEM }}/snowinc.{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
```
Should be `COMOUT`:
```suggestion
- ["{{ DATA }}/inc/ensmean/snowinc.{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ com_ns.COMOUT_SNOW_ANALYSIS_MEM }}/snowinc.{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
{% endfor %}
{% if DOIAU %}
# if using IAU, also need increment copied at the beginning of the window
{% for tile in range(1, 7) %}
- ["{{ DATA }}/inc/ensmean/snowinc.{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ com_ns.COMOUT_SNOW_ANALYSIS_MEM }}/snowinc.{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
```

DavidHuber-NOAA commented on parm/gdas/snow_finalize_ens_update.yaml.j2:
```diff
@@ -0,0 +1,70 @@
+{% if DOIAU %}
+# if using IAU, also need analyses copied at the beginning of the window
+{% for tile in range(1, 7) %}
+- ["{{ DATA }}/anl/mem{{ '%03d' % mem }}/{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ com_ns.COM_SNOW_ANALYSIS_MEM }}/{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
```
`COMOUT`
```suggestion
- ["{{ DATA }}/anl/mem{{ '%03d' % mem }}/{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ com_ns.COMOUT_SNOW_ANALYSIS_MEM }}/{{ SNOW_WINDOW_BEGIN | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
```

DavidHuber-NOAA commented on parm/gdas/snow_finalize_ens_update.yaml.j2:
```diff
@@ -0,0 +1,70 @@
+{% set com_ns.COM_SNOW_ANALYSIS_MEM = com_ns.COM_SNOW_ANALYSIS_MEM.replace(search_term, replace_term) %}
+{% endfor %}
+{% for tile in range(1, 7) %}
+- ["{{ DATA }}/anl/mem{{ '%03d' % mem }}/{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ com_ns.COM_SNOW_ANALYSIS_MEM }}/{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
```
`COMOUT`
```suggestion
{% set com_ns = namespace(COMOUT_SNOW_ANALYSIS_MEM = COMOUT_SNOW_ANALYSIS_TMPL) %}

{% for key in tmpl_dict.keys() %}
{% set search_term = '${' + key + '}' %}
{% set replace_term = tmpl_dict[key] %}
{% set com_ns.COMOUT_SNOW_ANALYSIS_MEM = com_ns.COMOUT_SNOW_ANALYSIS_MEM.replace(search_term, replace_term) %}
{% endfor %}
{% for tile in range(1, 7) %}
- ["{{ DATA }}/anl/mem{{ '%03d' % mem }}/{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc", "{{ com_ns.COMOUT_SNOW_ANALYSIS_MEM }}/{{ current_cycle | to_fv3time }}.sfc_data.tile{{ tile }}.nc"]
```

DavidHuber-NOAA commented on jobs/JGLOBAL_SNOW_ENSEMBLE_ANALYSIS:
```diff
@@ -0,0 +1,56 @@
+##############################################
+# Generate COM variables from templates
+YMD=${PDY} HH=${cyc} declare_from_tmpl -rx COM_OBS COM_ATMOS_ANALYSIS COM_CONF
+MEMDIR="ensstat" YMD=${PDY} HH=${cyc} declare_from_tmpl COM_SNOW_ANALYSIS
```
We should be moving to COMIN/COMOUT to indicate which directories are read from/written to:
```suggestion
YMD=${PDY} HH=${cyc} declare_from_tmpl -rx COMOUT_OBS COMOUT_ATMOS_ANALYSIS COMOUT_CONF
MEMDIR="ensstat" YMD=${PDY} HH=${cyc} declare_from_tmpl COMOUT_SNOW_ANALYSIS
```

DavidHuber-NOAA commented on jobs/JGLOBAL_SNOW_ENSEMBLE_ANALYSIS:
```diff
@@ -0,0 +1,56 @@
+
+for imem in $(seq 1 "${NMEM_ENS}"); do
+    memchar="mem$(printf %03i "${imem}")"
+    MEMDIR=${memchar} RUN=${GDUMP_ENS} YMD=${PDY} HH=${cyc} declare_from_tmpl COM_SNOW_ANALYSIS
```
```suggestion
    MEMDIR=${memchar} RUN=${GDUMP_ENS} YMD=${PDY} HH=${cyc} declare_from_tmpl COMOUT_SNOW_ANALYSIS
```

DavidHuber-NOAA commented on jobs/JGLOBAL_SNOW_ENSEMBLE_ANALYSIS:
```diff
@@ -0,0 +1,56 @@
+YMD=${PDY} HH=${cyc} declare_from_tmpl -rx COM_OBS COM_ATMOS_ANALYSIS COM_CONF
+MEMDIR="ensstat" YMD=${PDY} HH=${cyc} declare_from_tmpl COM_SNOW_ANALYSIS
+
+mkdir -m 775 -p "${COM_SNOW_ANALYSIS}" "${COM_CONF}"
```
```suggestion
mkdir -m 775 -p "${COMOUT_SNOW_ANALYSIS}" "${COMOUT_CONF}"
```

DavidHuber-NOAA commented on parm/gdas/snow_finalize_ens_update.yaml.j2:
```diff
@@ -0,0 +1,70 @@
+# This must be done in a namespace to overcome jinja scoping
+#  Variables set inside of a for loop are lost at the end of the loop
+#  unless they are part of a namespace
+{% set com_ns = namespace(COM_SNOW_ANALYSIS_MEM = COM_SNOW_ANALYSIS_TMPL) %}
+
+{% for key in tmpl_dict.keys() %}
+{% set search_term = '${' + key + '}' %}
+{% set replace_term = tmpl_dict[key] %}
+{% set com_ns.COM_SNOW_ANALYSIS_MEM = com_ns.COM_SNOW_ANALYSIS_MEM.replace(search_term, replace_term) %}
```
{% set com_ns = namespace(COMOUT_SNOW_ANALYSIS_MEM = COMOUT_SNOW_ANALYSIS_TMPL) %}

{% for key in tmpl_dict.keys() %}
{% set search_term = '${' + key + '}' %}
{% set replace_term = tmpl_dict[key] %}
{% set com_ns.COMOUT_SNOW_ANALYSIS_MEM = com_ns.COMOUT_SNOW_ANALYSIS_MEM.replace(search_term, replace_term) %}

DavidHuber-NOAA commented on scripts/exgdas_enkf_sfc.sh:
COM should be renamed to COMIN/COMOUT.

CoryMartin-NOAA replied:
Is there anyone else tackling this for the rest of this job (I'm only touching part of it) or should I make the COM changes for the entirety of this job in this PR?

DavidHuber-NOAA replied:
Hmm, good question.  Since this isn't touching the whole job, I'm OK with tackling this in a separate PR.

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
update gdas hash

CoryMartin-NOAA committed:
address reviewer comments

CoryMartin-NOAA commented:
Thanks @DavidHuber-NOAA and sorry for the delay in replying. I think I've addressed all comments except the ones in the existing enkf_sfc job, to which I have a question about.

CoryMartin-NOAA committed:
update name of test

DavidHuber-NOAA commented on env/WCOSS2.env:
```diff
@@ -85,7 +85,17 @@ elif [[ "${step}" = "snowanl" ]]; then
     export APRUN_SNOWANL="${launcher} -n ${npe_snowanl}"
 
-    export APRUN_APPLY_INCR="${launcher} -n 6"
+    export APRUN_APPLY_INCR="${launcher} -n ${npe_apply_incr}6"
```
```suggestion
    export APRUN_APPLY_INCR="${launcher} -n ${npe_apply_incr}"
```

CoryMartin-NOAA committed:
Update env/WCOSS2.env

Co-authored-by: David Huber <69919478+DavidHuber-NOAA@users.noreply.github.com>

aerorahul commented on ci/Jenkinsfile:
```diff
@@ -181,14 +181,14 @@ pipeline {
                         name 'Case'
                         // TODO add dynamic list of cases from env vars (needs addtional plugins)
-                        values 'C48C48_ufs_hybatmDA', 'C48_ATM', 'C48_S2SW', 'C48_S2SWA_gefs', 'C48mx500_3DVarAOWCDA', 'C96C48_hybatmDA', 'C96_atm3DVar', 'C96_atmaerosnowDA'
+                        values 'C48C48_ufs_hybatmDA', 'C48_ATM', 'C48_S2SW', 'C48_S2SWA_gefs', 'C48mx500_3DVarAOWCDA', 'C96C48_hybatmDA', 'C96_atm3DVar', 'C96_hybatmaerosnowDA'
```
```suggestion
                        values 'C48C48_ufs_hybatmDA', 'C48_ATM', 'C48_S2SW', 'C48_S2SWA_gefs', 'C48mx500_3DVarAOWCDA', 'C96C48_hybatmDA', 'C96_atm3DVar', 'C96C48_hybatmaerosnowDA'
```
```suggestion
                        values 'C48C48_ufs_hybatmDA', 'C48_ATM', 'C48_S2SW', 'C48_S2SWA_gefs', 'C48mx500_3DVarAOWCDA', 'C96C48_hybatmDA', 'C96_atm3DVar', 'C96_hybatmaerosnowDA'
```

aerorahul commented on ci/cases/pr/C96_hybatmaerosnowDA.yaml:
Rename to `C96C48_hybatmaerosnowDA.yaml` to include the ensemble resolution

aerorahul commented on jobs/JGLOBAL_SNOW_ENSEMBLE_ANALYSIS:
```diff
@@ -0,0 +1,56 @@
+# Ignore possible spelling error (nothing is misspelled)
+# shellcheck disable=SC2153
+GDUMP="gdas"
+GDUMP_ENS="enkf${GDUMP}"
```
```suggestion
GDUMP_ENS="enkfgdas"
```
Both are always going to be `gdas` and `enkfgdas`

aerorahul commented on jobs/JGLOBAL_SNOW_ENSEMBLE_ANALYSIS:
```diff
@@ -0,0 +1,56 @@
+# Begin JOB SPECIFIC work
+##############################################
+# Generate COM variables from templates
+YMD=${PDY} HH=${cyc} declare_from_tmpl -rx COMIN_OBS COMOUT_ATMOS_ANALYSIS COMOUT_CONF
+MEMDIR="ensstat" YMD=${PDY} HH=${cyc} declare_from_tmpl COMOUT_SNOW_ANALYSIS
```
These will fail, because there are not template with `COMIN_` or `COMOUT_`.  The templates are `COM_`.
```suggestion
YMD=${PDY} HH=${cyc} declare_from_tmpl -rx \
  COMIN_OBS:COM_OBS_TMPL \
  COMOUT_ATMOS_ANALYSIS:COM_ATMOS_ANALYSIS_TMPL \
  COMOUT_CONF: COM_CONF_TMPL
MEMDIR="ensstat" YMD=${PDY} HH=${cyc} declare_from_tmpl 
 COMOUT_SNOW_ANALYSIS:COM_SNOW_ANALYSIS_TMPL
```
Just a note that `COMOUT_SNOW_ANALYSIS` is not being exported, so it will stay local (i.e. child scripts `exscripts...` will not know this variable.

aerorahul commented on jobs/JGLOBAL_SNOW_ENSEMBLE_ANALYSIS:
```diff
@@ -0,0 +1,56 @@
+YMD=${PDY} HH=${cyc} declare_from_tmpl -rx COMIN_OBS COMOUT_ATMOS_ANALYSIS COMOUT_CONF
+MEMDIR="ensstat" YMD=${PDY} HH=${cyc} declare_from_tmpl COMOUT_SNOW_ANALYSIS
+
+mkdir -m 775 -p "${COMOUT_SNOW_ANALYSIS}" "${COMOUT_CONF}"
```
```suggestion
mkdir -p "${COMOUT_SNOW_ANALYSIS}" "${COMOUT_CONF}"
```
`-m 775` gives group write permissions to the directory.  Don't think we want that.

CoryMartin-NOAA replied:
I forget where I pulled this from, other jobs definitely do this, is there a reason why you think?

aerorahul replied:
yeah, we noticed it was done in the wave jobs.
I looked at the permissions, and it gives write permission to members belonging to the group.  I don't think that is right.  
ops runs in the ops/spa role account, so they don't need this.

aerorahul commented on jobs/JGLOBAL_SNOW_ENSEMBLE_ANALYSIS:
```diff
@@ -0,0 +1,56 @@
+
+for imem in $(seq 1 "${NMEM_ENS}"); do
+    memchar="mem$(printf %03i "${imem}")"
+    MEMDIR=${memchar} RUN=${GDUMP_ENS} YMD=${PDY} HH=${cyc} declare_from_tmpl COMOUT_SNOW_ANALYSIS
+    mkdir -m 775 -p "${COMOUT_SNOW_ANALYSIS}"
```
```suggestion
    MEMDIR=${memchar} RUN=${GDUMP_ENS} YMD=${PDY} HH=${cyc} declare_from_tmpl COMOUT_SNOW_ANALYSIS:COM_SNOW_ANALYSIS_TMPL
    mkdir -p "${COMOUT_SNOW_ANALYSIS}"
```
Noting here that `COMOUT_SNOW_ANALYSIS` will not be exported, just declared.

aerorahul commented on jobs/JGLOBAL_SNOW_ENSEMBLE_ANALYSIS:
```diff
@@ -0,0 +1,56 @@
+EXSCRIPT=${SNOWANLPY:-${SCRgfs}/exglobal_snow_ensemble_analysis.py}
+${EXSCRIPT}
+status=$?
+[[ ${status} -ne 0 ]] && exit "${status}"
```
```suggestion
(( status != 0 )) && exit "${status}"
```
just a suggestion here for future note.

aerorahul commented on jobs/JGLOBAL_SNOW_ENSEMBLE_ANALYSIS:
```diff
@@ -0,0 +1,56 @@
+# Remove the Temporary working directory
+##########################################
+cd "${DATAROOT}" || exit 1
+[[ ${KEEPDATA} = "NO" ]] && rm -rf "${DATA}"
```
```suggestion
[[ "${KEEPDATA}" = "NO" ]] && rm -rf "${DATA}"
```

aerorahul commented on jobs/rocoto/esnowanl.sh:
```diff
@@ -0,0 +1,24 @@
+export job="esnowanl"
+export jobid="${job}.$$"
+
+###############################################################
+# setup python path for workflow utilities and tasks
+wxflowPATH="${HOMEgfs}/ush/python:${HOMEgfs}/ush/python/wxflow/src"
+PYTHONPATH="${PYTHONPATH:+${PYTHONPATH}:}${wxflowPATH}"
+export PYTHONPATH
```
This will depend on the other PR that adds `wxflow` to `load_ufsda_modules.sh`.  That will greatly simplify things.
Just a note as this may or may not be needed based on time of merge

aerorahul replied:
```suggestion
```
#2700 moved this section to `load_ufsda_modules.sh`.  It is no longer needed in the rocoto job.

aerorahul commented on parm/config/gfs/config.esnowanl:
```diff
@@ -0,0 +1,29 @@
+export io_layout_y=@IO_LAYOUT_Y@
+
+export JEDIEXE=${EXECgfs}/gdasapp_land_ensrecenter.x
+export FREGRID=${EXECgfs}/fregrid.x
```
Just left a note on `fregrid.x` in ufs-utils re. compiler warnings and remarks in preparation for v17.

CoryMartin-NOAA replied:
I'll double check this and get back with you

aerorahul replied:
George Gayno has found and fixed the compile time warnings.

aerorahul commented on parm/config/gfs/config.resources:
```diff
@@ -369,6 +369,37 @@ case ${step} in
+  "esnowanl")
+    # below lines are for creating JEDI YAML
+    case ${CASE} in
+      "C768")
+        layout_x=6
+        layout_y=6
+        ;;
+      "C384")
+        layout_x=5
+        layout_y=5
+        ;;
```
Is using `C768` settings for `C384` too much?  Just curious.

CoryMartin-NOAA replied:
These need to be revisited once we've actually ran at higher resolutions.

aerorahul commented on parm/gdas/snow_finalize_ens_update.yaml.j2:
```diff
@@ -0,0 +1,70 @@
+    {% set com_ns = namespace(COMOUT_SNOW_ANALYSIS_MEM = COM_SNOW_ANALYSIS_TMPL) %}
+
+    {% for key in tmpl_dict.keys() %}
+        {% set search_term = '${' + key + '}' %}
+        {% set replace_term = tmpl_dict[key] %}
+        {% set com_ns.COMOUT_SNOW_ANALYSIS_MEM = com_ns.COMOUT_SNOW_ANALYSIS_MEM.replace(search_term, replace_term) %}
```
@DavidNew-NOAA added a nifty filter for this in wxflow.  Please see PR #2700

CoryMartin-NOAA replied:
I forgot. I still need to address this

aerorahul commented on parm/gdas/snow_finalize_ens_update.yaml.j2:
```diff
@@ -0,0 +1,70 @@
+        {% set replace_term = tmpl_dict[key] %}
+        {% set com_ns.COMOUT_SNOW_ANALYSIS_MEM = com_ns.COMOUT_SNOW_ANALYSIS_MEM.replace(search_term, replace_term) %}
+    {% endfor %}
+    {% for tile in range(1, 7) %}
```
```suggestion
    {% for tile in range(1, ntiles+1) %}
```

CoryMartin-NOAA committed:
no need for a non-standard DATA directory

aerorahul commented on parm/gdas/snow_finalize_ens_update.yaml.j2:
```diff
@@ -0,0 +1,70 @@
+    {% endfor %}
+    {% if DOIAU == True %}
+        # if using IAU, also need analyses copied at the beginning of the window
+        {% for tile in range(1, 7) %}
```
```suggestion
        {% for tile in range(1, ntiles+1) %}
```

aerorahul commented on parm/gdas/snow_finalize_ens_update.yaml.j2:
```diff
@@ -0,0 +1,70 @@
+{% for key in tmpl_dict.keys() %}
+    {% set search_term = '${' + key + '}' %}
+    {% set replace_term = tmpl_dict[key] %}
+    {% set com_ns.COMOUT_SNOW_ANALYSIS_MEM = com_ns.COMOUT_SNOW_ANALYSIS_MEM.replace(search_term, replace_term) %}
```
Same comment as above

aerorahul commented on parm/gdas/snow_finalize_ens_update.yaml.j2:
```diff
@@ -0,0 +1,70 @@
+{% endfor %}
+{% if DOIAU == True %}
+    # if using IAU, also need increment copied at the beginning of the window
+    {% for tile in range(1, 7) %}
```
```suggestion
    {% for tile in range(1, ntiles+1) %}
```

aerorahul commented on parm/gdas/snow_finalize_ens_update.yaml.j2:
```diff
@@ -0,0 +1,70 @@
+# set some variables
+######################################
+{% set PDY = current_cycle | to_YMD %}
+{% set cyc = current_cycle | strftime("%H") %}
```
I would not use `PDY` and `cyc` and use `current_cycle` and the appropriate filter where it is used.

aerorahul commented on parm/gdas/snow_stage_ens_update.yaml.j2:
```diff
@@ -0,0 +1,113 @@
+{% set gPDY = previous_cycle | to_YMD %}
+{% set gcyc = previous_cycle | strftime("%H") %}
+{% set PDY = current_cycle | to_YMD %}
+{% set cyc = current_cycle | strftime("%H") %}
```
I would use the `current_cycle` and `previous_cycle` with the appropriate filter where needed.

aerorahul commented on parm/gdas/snow_stage_ens_update.yaml.j2:
```diff
@@ -0,0 +1,113 @@
+    {% set replace_term = tmpl_dict[key] %}
+    {% set com_prev_ns.COMIN_ATMOS_RESTART_MEM = com_prev_ns.COMIN_ATMOS_RESTART_MEM.replace(search_term, replace_term) %}
+{% endfor %}
+{% for tile in range(1, 7) %}
```
```suggestion
{% for tile in range(1, ntiles+1) %}
```

aerorahul commented on parm/gdas/snow_stage_ens_update.yaml.j2:
```diff
@@ -0,0 +1,113 @@
+    {% set replace_term = tmpl_dict[key] %}
+    {% set com_ns.COMIN_SNOW_ANALYSIS_MEM = com_ns.COMIN_SNOW_ANALYSIS_MEM.replace(search_term, replace_term) %}
+{% endfor %}
+{% for tile in range(1, 7) %}
```
```suggestion
{% for tile in range(1, ntiles+1) %}
```

aerorahul commented on parm/gdas/snow_stage_ens_update.yaml.j2:
```diff
@@ -0,0 +1,113 @@
+        {% set com_prev_ns.COMIN_ATMOS_RESTART_MEM = com_prev_ns.COMIN_ATMOS_RESTART_MEM.replace(search_term, replace_term) %}
+    {% endfor %}
+    # we need to copy them to two places, one serves as the basis for the analysis
+    {% for tile in range(1, 7) %}
```
```suggestion
    {% for tile in range(1, ntiles+1) %}
```

aerorahul commented on parm/gdas/snow_stage_ens_update.yaml.j2:
```diff
@@ -0,0 +1,113 @@
+    {% if DOIAU == True %}
+        # if using IAU, also need backgrounds copied at the beginning of the window
+        # we need to copy them to two places, one serves as the basis for the analysis
+        {% for tile in range(1, 7) %}
```
```suggestion
        {% for tile in range(1, ntiles) %}
```

aerorahul commented on parm/gdas/snow_stage_orog.yaml.j2:
```diff
@@ -0,0 +1,12 @@
+- "{{ DATA }}/orog/det"
+- "{{ DATA }}/orog/ens"
+copy:
+- ["{{ FIXgfs }}/orog/{{ CASE }}/{{ CASE }}_mosaic.nc", "{{ DATA }}/orog/det/{{ CASE }}_mosaic.nc"]
+- ["{{ FIXgfs }}/orog/{{ CASE_ENS }}/{{ CASE_ENS }}_mosaic.nc", "{{ DATA }}/orog/ens/{{ CASE_ENS }}_mosaic.nc"]
+{% for tile in range(1, 7) %}
+- ["{{ FIXgfs }}/orog/{{ CASE }}/{{ CASE }}_grid.tile{{ tile }}.nc", "{{ DATA }}/orog/det/{{ CASE }}_grid.tile{{ tile }}.nc"]
+- ["{{ FIXgfs }}/orog/{{ CASE_ENS }}/{{ CASE_ENS }}_grid.tile{{ tile }}.nc", "{{ DATA }}/orog/ens/{{ CASE_ENS }}_grid.tile{{ tile }}.nc"]
+- ["{{ FIXgfs }}/orog/{{ CASE }}/{{ CASE }}.mx{{ OCNRES }}_oro_data.tile{{ tile }}.nc", "{{ DATA }}/orog/det/{{ CASE }}.mx{{ OCNRES }}_oro_data.tile{{ tile }}.nc" ]
+- ["{{ FIXgfs }}/orog/{{ CASE_ENS }}/{{ CASE_ENS }}.mx{{ OCNRES }}_oro_data.tile{{ tile }}.nc", "{{ DATA }}/orog/ens/{{ CASE_ENS }}.mx{{ OCNRES }}_oro_data.tile{{ tile }}.nc" ]
```
The nest case required us to define `FIXorog` instead of using `FIXgfs/orog`.

aerorahul commented on parm/gdas/snow_stage_orog.yaml.j2:
```diff
@@ -0,0 +1,12 @@
+copy:
+- ["{{ FIXgfs }}/orog/{{ CASE }}/{{ CASE }}_mosaic.nc", "{{ DATA }}/orog/det/{{ CASE }}_mosaic.nc"]
+- ["{{ FIXgfs }}/orog/{{ CASE_ENS }}/{{ CASE_ENS }}_mosaic.nc", "{{ DATA }}/orog/ens/{{ CASE_ENS }}_mosaic.nc"]
+{% for tile in range(1, 7) %}
```
```suggestion
{% for tile in range(1, 7) %}
```
```suggestion
{% for tile in range(1, ntiles) %}
```

jiaruidong2017 replied:
Should be:    `{% for tile in range(1, ntiles+1) %}`

aerorahul commented on scripts/exgdas_enkf_sfc.sh:
```diff
@@ -193,37 +229,60 @@ if [ $DOSFCANL_ENKF = "YES" ]; then
 
-            MEMDIR=${memchar} YMD=${PDY} HH=${cyc} declare_from_tmpl \
-                COM_ATMOS_RESTART_MEM:COM_ATMOS_RESTART_TMPL
+            COM_SNOW_ANALYSIS_TMPL
```
Is this a stray string here?

DavidHuber-NOAA commented on scripts/exgdas_enkf_sfc.sh:
```diff
@@ -193,37 +229,60 @@ if [ $DOSFCANL_ENKF = "YES" ]; then
 
-            MEMDIR=${memchar} YMD=${PDY} HH=${cyc} declare_from_tmpl \
-                COM_ATMOS_RESTART_MEM:COM_ATMOS_RESTART_TMPL
+            COM_SNOW_ANALYSIS_TMPL
+
```
```suggestion
```

CoryMartin-NOAA replied:
Whoops!

CoryMartin-NOAA committed:
Update scripts/exgdas_enkf_sfc.sh

Co-authored-by: David Huber <69919478+DavidHuber-NOAA@users.noreply.github.com>

aerorahul commented on scripts/exgdas_enkf_sfc.sh:
```diff
@@ -193,37 +229,60 @@ if [ $DOSFCANL_ENKF = "YES" ]; then
-            ${NLN} "${FIXorog}/${CASE}/${CASE}_grid.tile${n}.nc"      "${DATA}/fngrid.${cmem}"
-            ${NLN} "${FIXorog}/${CASE}/${CASE}.mx${OCNRES}_oro_data.tile${n}.nc" "${DATA}/fnorog.${cmem}"
+            ${NCP} "${DATA}/fnbgsi.${cmem}" "${DATA}/fnbgso.${cmem}"
+            ${NLN} "${FIXgfs}/orog/${CASE}/${CASE}_grid.tile${n}.nc"      "${DATA}/fngrid.${cmem}"
+            ${NLN} "${FIXgfs}/orog/${CASE}/${CASE}.mx${OCNRES}_oro_data.tile${n}.nc" "${DATA}/fnorog.${cmem}"
```
EE2 wants copy instead of link.  
We did that in the deterministic script.

aerorahul commented on scripts/exgdas_enkf_sfc.sh:
```diff
@@ -175,6 +183,33 @@ if [ $DOIAU = "YES" ]; then
+
+            if [[ ${GSI_SOILANAL} = "YES" ]]; then
+                FHR=6
+                ${NLN} "${COM_ATMOS_ANALYSIS_MEM}/${APREFIX_ENS}sfci00${FHR}.nc" \
+                   "${DATA}/lnd_incr.${cmem}"
```
No linking to COM.  Please copy

aerorahul commented on scripts/exgdas_enkf_sfc.sh:
```diff
@@ -150,20 +151,27 @@ if [ $DOIAU = "YES" ]; then
-            ${NLN} "${FIXorog}/${CASE}/${CASE}_grid.tile${n}.nc"     "${DATA}/fngrid.${cmem}"
-            ${NLN} "${FIXorog}/${CASE}/${CASE}.mx${OCNRES}_oro_data.tile${n}.nc" "${DATA}/fnorog.${cmem}"
+            ${NCP} "${DATA}/fnbgsi.${cmem}" "${DATA}/fnbgso.${cmem}"
+            ${NLN} "${FIXgfs}/orog/${CASE}/${CASE}_grid.tile${n}.nc"     "${DATA}/fngrid.${cmem}"
+            ${NLN} "${FIXgfs}/orog/${CASE}/${CASE}.mx${OCNRES}_oro_data.tile${n}.nc" "${DATA}/fnorog.${cmem}"
```
Needs to be copied, not linked.

DavidHuber-NOAA reviewed (dismissed):
Looks good to me.

aerorahul reviewed (commented):
first pass of review.

DavidHuber-NOAA commented:
@CoryMartin-NOAA Could you also request a review from someone with more knowledge of snow DA?  Perhaps @jiaruidong2017?

CoryMartin-NOAA commented:
Done. @jiaruidong2017 please review. Any issues we find/solve in terms of the science we can fix in a future PR, for now, just review the 'technical' aspects here.

CoryMartin-NOAA committed:
address reviewer comments

jiaruidong2017 commented:
@CoryMartin-NOAA I conducted another testrun after I updated your branch and rebuilt the package. However, it failed with the following errors:
```
+ JGLOBAL_SNOW_ENSEMBLE_ANALYSIS[23]: MEMDIR=ensstat
+ JGLOBAL_SNOW_ENSEMBLE_ANALYSIS[23]: YMD=20211220
+ JGLOBAL_SNOW_ENSEMBLE_ANALYSIS[23]: HH=18
+ JGLOBAL_SNOW_ENSEMBLE_ANALYSIS[23]: declare_from_tmpl ' '
+ bash_utils.sh[35]: [[ NO == \N\O ]]
+ bash_utils.sh[35]: set +x
FATAL ERROR in declare_from_tmpl: Requested template  _TMPL not defined!
End JGLOBAL_SNOW_ENSEMBLE_ANALYSIS at 03:30:30 with error code 2 (time elapsed: 00:00:16)
+ esnowanl.sh[1]: postamble esnowanl.sh 1719718187 2
+ preamble.sh[70]: set +x
End esnowanl.sh at 03:30:30 with error code 2 (time elapsed: 00:00:43)
```
I checked the code `jobs/JGLOBAL_SNOW_ENSEMBLE_ANALYSIS` as below: 
```
MEMDIR="ensstat" YMD=${PDY} HH=${cyc} declare_from_tmpl \
    COMOUT_SNOW_ANALYSIS:COM_SNOW_ANALYSIS_TMPL
```
The requested template `_TMPL` was defined there. I don't know what caused this issue. 

In addition, I re-cloned, rebuilt the package and re-run another test, it still failed with the same error messages. The whole logfile on hera is at: `/scratch1/NCEPDEV/climate/Jiarui.Dong/ptmp/cory04/logs/2021122018/enkfgdasesnowanl.log`

CoryMartin-NOAA commented:
Thanks @jiaruidong2017 I think this is something that may have gotten broken during one of the rounds of addressing reviewer comments. I'll take a look later today.

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

jiaruidong2017 commented on jobs/JGLOBAL_SNOW_ENSEMBLE_ANALYSIS:
```diff
@@ -0,0 +1,59 @@
+
+for imem in $(seq 1 "${NMEM_ENS}"); do
+    memchar="mem$(printf %03i "${imem}")"
+    MEMDIR=${memchar} RUN=${GDUMP_ENS} YMD=${PDY} HH=${cyc} declare_from_tmpl COMOUT_SNOW_ANALYSIS
```
Change to:  
```
     MEMDIR=${memchar} RUN=${GDUMP_ENS} YMD=${PDY} HH=${cyc} declare_from_tmpl \
        COMOUT_SNOW_ANALYSIS:COM_SNOW_ANALYSIS_TMPL
```

CoryMartin-NOAA committed:
Merge branch 'feature/snow-ensemble' of http://github.com/corymartin-noaa/global-workflow into feature/snow-ensemble

jiaruidong2017 commented on jobs/JGLOBAL_SNOW_ENSEMBLE_ANALYSIS:
```diff
@@ -0,0 +1,59 @@
+    COMIN_OBS:COM_OBS_TMPL \
+    COMOUT_ATMOS_ANALYSIS:COM_ATMOS_ANALYSIS_TMPL \
+    COMOUT_CONF:COM_CONF_TMPL 
+MEMDIR="ensstat" YMD=${PDY} HH=${cyc} declare_from_tmpl \ 
```
Delete the space after `\`

CoryMartin-NOAA committed:
fix template issue + updates for g-w develop merge

CoryMartin-NOAA commented:
@jiaruidong2017 I think I've fixed the issue you found (and some others), try it now.

jiaruidong2017 reviewed (approved):
Looks good to me. Thanks

CoryMartin-NOAA committed:
Address Jiarui's comments

DavidHuber-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

DavidHuber-NOAA commented:
I may have updated with develop too soon.  I see there are two comments from @aerorahul regarding copying rather than linking to `COM`.  Once these are addressed, I think this PR is ready for CI testing.

CoryMartin-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
No NLN but NCP

DavidHuber-NOAA reviewed (dismissed):
Approved pending successful CI testing.

aerorahul commented on jobs/JGLOBAL_SNOW_ENSEMBLE_ANALYSIS:
```diff
@@ -0,0 +1,60 @@
+
+for imem in $(seq 1 "${NMEM_ENS}"); do
+    memchar="mem$(printf %03i "${imem}")"
+    MEMDIR=${memchar} RUN=${GDUMP_ENS} YMD=${PDY} HH=${cyc} declare_from_tmpl \
+        COMOUT_SNOW_ANALYSIS:COM_SNOW_ANALYSIS_TMPL
```
`GDUMP_ENS=gdas`.
When this script is run with `RUN=enkfgfs`, this will create a variable in the `enkfgdas` path for `COMOUT`.
Is that correct?
Do we plan to execute `JGLOBAL_SNOW_ENSEMBLE_ANALYSIS` in the early cycle with `RUN=enkfgfs`?

CoryMartin-NOAA replied:
I think you and I chatted about this previously, as we don't run an early cycle enkfgfs fcst, there's no need to do this, the GEFS can decide if they wish to do this recentering?

aerorahul replied:
In that case, we need to name this job `JENKFGDAS_` instead of `JGLOBAL_` to conform with `JRUN_` convention.

CoryMartin-NOAA replied:
is it `JGDAS_ENKF` to be consistent with the others or `JENKFGDAS_` ?

CoryMartin-NOAA committed:
Update jobs/rocoto/esnowanl.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

aerorahul commented on scripts/exgdas_enkf_sfc.sh:
```diff
@@ -150,31 +151,65 @@ if [ $DOIAU = "YES" ]; then
-            ${NLN} "${FIXorog}/${CASE}/${CASE}_grid.tile${n}.nc"     "${DATA}/fngrid.${cmem}"
-            ${NLN} "${FIXorog}/${CASE}/${CASE}.mx${OCNRES}_oro_data.tile${n}.nc" "${DATA}/fnorog.${cmem}"
+            ${NCP} "${DATA}/fnbgsi.${cmem}" "${DATA}/fnbgso.${cmem}"
+            ${NCP} "${FIXgfs}/orog/${CASE}/${CASE}_grid.tile${n}.nc"     "${DATA}/fngrid.${cmem}"
```
```suggestion
            ${NCP} "${FIXorog}/${CASE}/${CASE}_grid.tile${n}.nc"     "${DATA}/fngrid.${cmem}"
```

aerorahul commented on scripts/exgdas_enkf_sfc.sh:
```diff
@@ -150,31 +151,65 @@ if [ $DOIAU = "YES" ]; then
-            ${NLN} "${FIXorog}/${CASE}/${CASE}.mx${OCNRES}_oro_data.tile${n}.nc" "${DATA}/fnorog.${cmem}"
+            ${NCP} "${DATA}/fnbgsi.${cmem}" "${DATA}/fnbgso.${cmem}"
+            ${NCP} "${FIXgfs}/orog/${CASE}/${CASE}_grid.tile${n}.nc"     "${DATA}/fngrid.${cmem}"
+            ${NCP} "${FIXgfs}/orog/${CASE}/${CASE}.mx${OCNRES}_oro_data.tile${n}.nc" "${DATA}/fnorog.${cmem}"
```
```suggestion
            ${NCP} "${FIXorog}/${CASE}/${CASE}.mx${OCNRES}_oro_data.tile${n}.nc" "${DATA}/fnorog.${cmem}"
```

aerorahul commented on scripts/exglobal_snow_ensemble_analysis.py:
```diff
@@ -0,0 +1,27 @@
+import os
+
+from wxflow import Logger, cast_strdict_as_dtypedict
+from pygfs.task.snowens_analysis import SnowEnsAnalysis
```
```suggestion
from pygfs.snowens_analysis import SnowEnsAnalysis
```
I think this should work. 
The others will be updated as well.

aerorahul commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,371 @@
+                    logit,
+                    Executable,
+                    WorkflowException)
+from pygfs.task.analysis import Analysis
```
```suggestion
from pygfs.analysis import Analysis
```

CoryMartin-NOAA replied:
I'm not sure why (yet) but while the first suggestion above worked, this one does not seem to.

CoryMartin-NOAA replied:
Never mind, neither work, I'm going to not remove `task` until this is sorted out, I think it's outside the scope of this PR.

aerorahul commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,371 @@
+            template_in = f'snowinc.{to_fv3time(self.task_config.SNOW_WINDOW_BEGIN)}.sfc_data.tile{{tilenum}}.nc'
+            template_out = f'snowinc.{to_fv3time(self.task_config.current_cycle)}.sfc_data.tile{{tilenum}}.nc'
+            inclist = []
+            for itile in range(1, 7):
```
```suggestion
            for itile in range(1, self.task_config.ntiles + 1):
```

aerorahul commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,371 @@
+                        'current_cycle': bkg_time,
+                        'CASE_ENS': self.task_config.CASE_ENS,
+                        'OCNRES': self.task_config.OCNRES,
+                        'ntiles': 6,
```
```suggestion
                        'ntiles': self.task_config.ntiles,
```

aerorahul commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,371 @@
+        WorkflowException
+            All other exceptions
+        """
+        os.chdir(config.DATA)
```
Using `os.chdir` here.  Seen `wxflow` `chdir` elsewhere in this script.  Just FYI.

CoryMartin-NOAA replied:
I don't remember the details, but there was a reason why I did this. For whatever reason, `wxflow chdir` was not working and it didn't `cd` into each subdirectory, but this did work.

aerorahul commented on workflow/applications/gfs_cycled.py:
```diff
@@ -167,6 +169,8 @@ def get_task_names(self):
             else:
                 hybrid_tasks += ['eobs', 'eupd', 'echgres']
                 hybrid_tasks += ['ediag'] if self.lobsdiag_forenkf else ['eomg']
+            if self.do_jedisnowda:
+                hybrid_tasks += ['esnowanl']
```
Just curious if one can run JEDI SNOW DA in the deterministic only, but not in the ensemble when running in hybrid.

CoryMartin-NOAA replied:
That's kind of what we are doing here, all this job does is recenter the ensemble to the deterministic, we don't have an EDA for the snow.

aerorahul commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,371 @@
+        os.symlink(exe_src, exe_dest)
+
+        # execute APPLY_INCR_EXE to create analysis files
+        print(os.getcwd())
```
```suggestion
```

aerorahul commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,371 @@
+import os
+from logging import getLogger
+from typing import Dict, List, Any
+from pprint import pformat
```
Check if this is used

aerorahul commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,371 @@
+
+from wxflow import (AttrDict,
+                    FileHandler,
+                    to_fv3time, to_YMD, to_YMDH, to_timedelta, add_to_datetime,
```
Check if `to_YMDH` is used

jiaruidong2017 replied:
In sorc/gdas.cd/parm/snow/prep/prep_ims.yaml.j2:

ims2ioda:
```
  copy:
    - ['{{ DATA }}/ims_snow_{{ current_cycle | to_YMDH }}.nc4', '{{ COM_OBS }}/{{ OPREFIX }}ims_snow.nc4']

```

aerorahul replied:
I should have said specifically in the python script.
It doesn't matter in the jinja-yaml as the filters are imported locally in `jinja2.py`.

aerorahul commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,371 @@
+from typing import Dict, List, Any
+from pprint import pformat
+import numpy as np
+from netCDF4 import Dataset
```
Check if Dataset is used

jiaruidong2017 replied:
This is not used

jiaruidong2017 commented:
This failure was caused by `mpiexec: unrecognized option '--cpus-per-task=1'` on wcoss2 as below:
```
[38;5;39m2024-07-03 18:51:16,215 - DEBUG    - snowens_analysis: Executing mpiexec -l -n 6 --cpus-per-task=1 /lfs/h2/emc/stmp/terry.mcguinness/RUNDIRS/C96C48_hybatmaerosnowDA_e2979b17/esnowanl.88786/gdasapp_land_ensrecenter.x /lfs/h2/emc/stmp/terry.mcguinness/RUNDIRS/C96C48_hybatmaerosnowDA_e2979b17/esnowanl.88786/enkfgdas.t18z.land_recenter.yaml[0m
mpiexec: unrecognized option '--cpus-per-task=1'
Usage: mpiexec [global_options] [command_options] cmd1
    [: [command_options] cmd2 [: ...] ]

```

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA commented:
The error that Jiarui pointed out should also be an issue in develop for the deterministic snow analysis. Why is it just now showing up?

CoryMartin-NOAA committed:
address comments

CoryMartin-NOAA committed:
Merge branch 'feature/snow-ensemble' of https://github.com/corymartin-noaa/global-workflow into feature/snow-ensemble

jiaruidong2017 commented:
> The error that Jiarui pointed out should also be an issue in develop for the deterministic snow analysis. Why is it just now showing up?

@CoryMartin-NOAA In gdas deterministic analysis, the `'--cpus-per-task=1'` option is not used in mpiexec there.

CoryMartin-NOAA committed:
Update WCOSS2.env

CoryMartin-NOAA commented:
> > The error that Jiarui pointed out should also be an issue in develop for the deterministic snow analysis. Why is it just now showing up?
> 
> @CoryMartin-NOAA In gdas deterministic analysis, the `'--cpus-per-task=1'` option is not used in mpiexec there.

Thanks @jiaruidong2017 yes I see what the issue is now, it should be fixed.

CoryMartin-NOAA committed:
use new wxflow technique for templates

CoryMartin-NOAA committed:
Merge branch 'feature/snow-ensemble' of https://github.com/corymartin-noaa/global-workflow into feature/snow-ensemble

CoryMartin-NOAA committed:
some python issues

CoryMartin-NOAA committed:
more issues

CoryMartin-NOAA committed:
fix templating

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA commented:
@DavidHuber-NOAA @aerorahul I think this is ready for review / testing but please correct me if I'm wrong

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
Merge branch 'feature/snow-ensemble' of https://github.com/corymartin-noaa/global-workflow into feature/snow-ensemble

WalterKolczynski-NOAA commented on scripts/exgdas_enkf_sfc.sh:
```diff
@@ -150,31 +151,65 @@ if [ $DOIAU = "YES" ]; then
+                COM_ATMOS_ANALYSIS_MEM:COM_ATMOS_ANALYSIS_TMPL
+
+            [[ ${TILE_NUM} -eq 1 ]] && mkdir -p "${COM_ATMOS_RESTART_MEM}"
+            ${NCP} "${DATA}/fnbgso.${cmem}" "${COM_ATMOS_RESTART_MEM}/${bPDY}.${bcyc}0000.sfcanl_data.tile${n}.nc"
```
Recommend all copies to COM use `cpfs` (except maybe small files) to avoid downstream jobs proceeding with partial files.

CoryMartin-NOAA replied:
Do you have an example of the preferred use of this that you can point me to?

WalterKolczynski-NOAA replied:
It's from `prod_util`. It just replaces the `$NCP` and copies the file to a temp name before then doing a rename (renaming is nearly instantaneous), so it prevents the use of a partial file before it has been fully written (or if the writing aborts in the middle for some reason). NCO recommends it for all large files going to COM. See p8 of the [implementation standards](https://www.nco.ncep.noaa.gov/idsb/implementation_standards/ImplementationStandards.v11.0.0.pdf).

WalterKolczynski-NOAA replied:
```suggestion
            cpfs "${DATA}/fnbgso.${cmem}" "${COM_ATMOS_RESTART_MEM}/${bPDY}.${bcyc}0000.sfcanl_data.tile${n}.nc"
```

CoryMartin-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/snow-ensemble

WalterKolczynski-NOAA commented on scripts/exgdas_enkf_sfc.sh:
```diff
@@ -193,37 +229,58 @@ if [ $DOSFCANL_ENKF = "YES" ]; then
+            MEMDIR=${memchar} YMD=${PDY} HH=${cyc} declare_from_tmpl \
+                COM_ATMOS_RESTART_MEM:COM_ATMOS_RESTART_TMPL
+
+            [[ ${TILE_NUM} -eq 1 ]] && mkdir -p "${COM_ATMOS_RESTART_MEM}"
```
```suggestion
            [[ ! -d "${COM_ATMOS_RESTART_MEM}" ]] && mkdir -p "${COM_ATMOS_RESTART_MEM}"
```

WalterKolczynski-NOAA commented on scripts/exgdas_enkf_sfc.sh:
```diff
@@ -193,37 +229,58 @@ if [ $DOSFCANL_ENKF = "YES" ]; then
 # Postprocessing
-cd $pwd
-[[ $mkdata = "YES" ]] && rm -rf $DATA
+cd "${pwd}" || exit 1
+[[ ${mkdata} = "YES" ]] && rm -rf "${DATA}"
```
ex-scripts should not be deleting `$DATA`; the j-job already does that.

CoryMartin-NOAA replied:
Yeah I think this was already here, I'll remove but note it just in case it breaks something

WalterKolczynski-NOAA commented on jobs/JGDAS_ENKF_SNOW_RECENTER:
```diff
@@ -0,0 +1,60 @@
+# shellcheck disable=SC2153
+GDUMP="gdas"
+GDUMP_ENS="enkfgdas"
+export GDUMP
```
Does `GDUMP_ENS` not need to be exported as well?

CoryMartin-NOAA replied:
This only runs for the GDAS so `RUN` is used in lieu of `GDUMP_ENS` in other places. I guess we can replace `GDUMP_ENS` in the j-job with `RUN`

CoryMartin-NOAA committed:
address some comments

CoryMartin-NOAA committed:
Merge branch 'feature/snow-ensemble' of https://github.com/corymartin-noaa/global-workflow into feature/snow-ensemble

CoryMartin-NOAA committed:
update GDAS hash

aerorahul commented on scripts/exgdas_enkf_sfc.sh:
```diff
@@ -193,37 +229,57 @@ if [ $DOSFCANL_ENKF = "YES" ]; then
+
+            [[ ! -d "${COM_ATMOS_RESTART_MEM}" ]] && mkdir -p "${COM_ATMOS_RESTART_MEM}"
+
+            ${NCP} "${DATA}/fnbgso.${cmem}" "${COM_ATMOS_RESTART_MEM}/${PDY}.${cyc}0000.sfcanl_data.tile${n}.nc"
```
Should this also be `cpfs`?

aerorahul commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,368 @@
+        except Exception:
+            raise WorkflowException(f"An error occured during execution of {exec_cmd}")
+
+        pass
+
```
```suggestion
```

CoryMartin-NOAA committed:
you shall not pass

aerorahul commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,368 @@
+            f"--weight_field land_frac",
+        ]
+        fregrid = os.path.join(self.task_config.DATA, 'fregrid.x') + " " + " ".join(arg_list)
+        exec_cmd = Executable(fregrid)
```
The executable is `fregrid.x` and that should be the `exec_cmd`.  The arguments should be passed to the `exec_cmd` via argument e.g. `exec_cmd(arg_list)` 
I can double check on the `__call__` in `wxflow.Executable`.

CoryMartin-NOAA replied:
yeah IIRC I had a lot of trouble getting this to work, so any help/suggestions to try would be appreciated. I know this works, most other things I tried the executable would complain about the arguments being passed to it

aerorahul replied:
ok. let me create a test example for you.

aerorahul replied:
```python3
fregrid_exe = os.path.join(self.task_config.DATA, 'fregrid.x')
exec_cmd = Executable(fregrid_exe)
exec_cmd(*arg_list)
```
Also added a test for this in wxflow so it serves as a reference for future use.
Please see [here](https://github.com/NOAA-EMC/wxflow/pull/38)

CoryMartin-NOAA replied:
Thanks I'll give this a try later today/tomorrow

CoryMartin-NOAA replied:
@aerorahul this did not work for me, I'm still getting the issues I was having previously...
```
/scratch1/NCEPDEV/stmp2/Cory.R.Martin/RUNDIRS/snowhighres/esnowrecen.2203935/fregrid.x: unrecognized option '--input_mosaic ./orog/det/C384_mosaic.nc'
/scratch1/NCEPDEV/stmp2/Cory.R.Martin/RUNDIRS/snowhighres/esnowrecen.2203935/fregrid.x: unrecognized option '--input_dir ./bkg/det/'
/scratch1/NCEPDEV/stmp2/Cory.R.Martin/RUNDIRS/snowhighres/esnowrecen.2203935/fregrid.x: unrecognized option '--input_file 20211220.150000.sfc_data'
/scratch1/NCEPDEV/stmp2/Cory.R.Martin/RUNDIRS/snowhighres/esnowrecen.2203935/fregrid.x: unrecognized option '--scalar_field snodl'
/scratch1/NCEPDEV/stmp2/Cory.R.Martin/RUNDIRS/snowhighres/esnowrecen.2203935/fregrid.x: unrecognized option '--output_dir ./bkg/det_ensres/'
/scratch1/NCEPDEV/stmp2/Cory.R.Martin/RUNDIRS/snowhighres/esnowrecen.2203935/fregrid.x: unrecognized option '--output_file 20211220.150000.sfc_data'
/scratch1/NCEPDEV/stmp2/Cory.R.Martin/RUNDIRS/snowhighres/esnowrecen.2203935/fregrid.x: unrecognized option '--output_mosaic ./orog/ens/C192_mosaic.nc'
/scratch1/NCEPDEV/stmp2/Cory.R.Martin/RUNDIRS/snowhighres/esnowrecen.2203935/fregrid.x: unrecognized option '--interp_method conserve_order1'
/scratch1/NCEPDEV/stmp2/Cory.R.Martin/RUNDIRS/snowhighres/esnowrecen.2203935/fregrid.x: unrecognized option '--weight_file ./orog/det/C384.mx025_interp_weight'
/scratch1/NCEPDEV/stmp2/Cory.R.Martin/RUNDIRS/snowhighres/esnowrecen.2203935/fregrid.x: unrecognized option '--weight_field lsm_frac'
/scratch1/NCEPDEV/stmp2/Cory.R.Martin/RUNDIRS/snowhighres/esnowrecen.2203935/fregrid.x: unrecognized option '--remap_file ./remap'
```

aerorahul commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,368 @@
+            f"--weight_field land_frac",
+        ]
+        fregrid = os.path.join(self.task_config.DATA, 'fregrid.x') + " " + " ".join(arg_list)
+        exec_cmd = Executable(fregrid)
```
Same comment on the use of `Executable` and passing arguments to the `exec_cmd` object.

CoryMartin-NOAA committed:
just remove mkdata stuff

CoryMartin-NOAA committed:
some updates

WalterKolczynski-NOAA reviewed (dismissed):
Conditionally approved pending successful completion of CI tests.

CoryMartin-NOAA commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,403 @@
+
+
+    @logit(logger)
+    def regridDetBkg(self) -> None:
```
@ClaraDraper-NOAA here is where the regridding takes place for the background

ClaraDraper-NOAA commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,403 @@
+            f"--input_mosaic ./orog/det/{self.task_config.CASE}_mosaic.nc",
+            f"--input_dir ./bkg/det/",
+            f"--input_file {to_fv3time(self.task_config.bkg_time)}.sfc_data",
+            f"--scalar_field snodl,slmsk,vtype",
```
Remove slmsk and vtype from here - they are integer fields, and will not be properly interpolated by fregrid.

ClaraDraper-NOAA commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,403 @@
+            f"--output_mosaic ./orog/ens/{self.task_config.CASE_ENS}_mosaic.nc",
+            f"--interp_method conserve_order1",
+            f"--weight_file ./orog/det/{self.task_config.CASE}.mx{self.task_config.OCNRES}_interp_weight",
+            f"--weight_field land_frac",
```
The "land_frac" in this file is land_frac with glaciers set to 0 to give the fraction of non-glacier land. To make this clearer, I recommend renaming it "lsm_frac", "soil_frac", 'nonglacier_land_frac"?

ClaraDraper-NOAA commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,403 @@
+            f"--output_mosaic ./orog/ens/{self.task_config.CASE_ENS}_mosaic.nc",
+            f"--interp_method conserve_order1",
+            f"--weight_file ./orog/det/{self.task_config.CASE}.mx{self.task_config.OCNRES}_interp_weight",
+            f"--weight_field land_frac",
```
same comment as above.

ClaraDraper-NOAA commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,403 @@
+        FileHandler(fregrid_copy).sync()
+
+    @logit(logger)
+    def genWeights(self) -> None:
```
Comment above, re naming.

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
address pynorms

CoryMartin-NOAA committed:
Address clara's comments + fregrid_exe

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

aerorahul commented:
@CoryMartin-NOAA 
Can you provide a log file for [this comment](https://github.com/NOAA-EMC/global-workflow/pull/2690#discussion_r1700139732)
Thanks!

CoryMartin-NOAA commented:
@aerorahul please see `/scratch1/NCEPDEV/da/Cory.R.Martin/work_july2024/snowens/com/snowhighres/logs/2021122018/enkfgdasesnowrecen.log.0`

aerorahul commented:
> @aerorahul please see `/scratch1/NCEPDEV/da/Cory.R.Martin/work_july2024/snowens/com/snowhighres/logs/2021122018/enkfgdasesnowrecen.log.0`

Thanks!
My gut says the issue is in the `arg_list` creation.
For e.g. `'--input_mosaic ./o
rog/det/C384_mosaic.nc'` should be 2 arguments `--input_mosaic`, `'./o
rog/det/C384_mosaic.nc'`

Is there a way I can try a standalone version of this executable to test this theory?

CoryMartin-NOAA commented:
@aerorahul there's `DATA=/scratch1/NCEPDEV/stmp2/Cory.R.Martin/RUNDIRS/snowhighres/esnowrecen.2203935` that is all set up as a test case from that attempt

aerorahul commented:
> @aerorahul there's `DATA=/scratch1/NCEPDEV/stmp2/Cory.R.Martin/RUNDIRS/snowhighres/esnowrecen.2203935` that is all set up as a test case from that attempt

![Screenshot 2024-08-01 at 11 15 26 AM](https://github.com/user-attachments/assets/b43370d6-6aba-4ca5-9692-4742d0972581)

CoryMartin-NOAA committed:
shellcheck

CoryMartin-NOAA commented:
@aerorahul your screenshot is missing a `5` at the end of the path...

CoryMartin-NOAA committed:
update from cdump to run

aerorahul commented:
> @aerorahul there's `DATA=/scratch1/NCEPDEV/stmp2/Cory.R.Martin/RUNDIRS/snowhighres/esnowrecen.2203935` that is all set up as a test case from that attempt

I setup a test case using this `DATA`
Please see `/scratch1/NCEPDEV/stmp2/Rahul.Mahajan/scratch/test_fregrid`. `test_fregrid.sh` drives `fregrid.py`.  `DATA` is copied from the above path.

`fregrid.stdout` contains the output from the executable.

CoryMartin-NOAA committed:
update fregrid arg lists

aerorahul commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -182,20 +183,21 @@ def regridDetInc(self) -> None:
-        fregrid_exe = os.path.join(self.task_config.DATA, 'fregrid.x') + " " + " ".join(arg_list)
+        fregrid_exe = os.path.join(self.task_config.DATA, 'fregrid.x')
         exec_cmd = Executable(fregrid_exe)
+        exec_cmd(*arg_list)
```
This needs to be removed and line 204 needs to be updated with this.

aerorahul commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -145,20 +145,21 @@ def regridDetBkg(self) -> None:
-        fregrid_exe = os.path.join(self.task_config.DATA, 'fregrid.x') + " " + " ".join(arg_list)
+        fregrid_exe = os.path.join(self.task_config.DATA, 'fregrid.x')
         exec_cmd = Executable(fregrid_exe)
+        exec_cmd(*arg_list)
```
This needs to be removed and line 166 should be updated instead.

CoryMartin-NOAA replied:
roger

CoryMartin-NOAA committed:
fix comments

WalterKolczynski-NOAA commented:
Is this ready to test again?

aerorahul reviewed (dismissed):
lgtm.
thanks for addressing the review comments

aerorahul committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA commented:
@WalterKolczynski-NOAA unfortunately no, something got broken as part of merging in all those resource changes. I'll fix it and let you know when it's ready for CI after my own testing.

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
Fix npe_apply_inc issue

CoryMartin-NOAA committed:
fix missing quote

CoryMartin-NOAA committed:
remove CDUMP

CoryMartin-NOAA committed:
update gdas hash

CoryMartin-NOAA commented:
@WalterKolczynski-NOAA I think we are ready to test now

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA commented:
@WalterKolczynski-NOAA I take it back, I am seeing some scientifically screwy results. I'm going to put this in draft until I'm ready for final review and testing.

CoryMartin-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

CoryMartin-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
update increment mask

CoryMartin-NOAA commented:
@WalterKolczynski-NOAA I think this is ready for final review and CI testing (hopefully)

guillaumevernieres committed:
Marine DA updates (#2802)

Companion PR to GDASApp#1226

Addresses a few issues described in:

- GDASApp#1219
- GDASApp#1217
- GDASApp#1232

AntonMFernando-NOAA committed:
Feature/get arch adds an empty archive job to GEFS system (#2772)

- This feature added an archive task for GEFS system. This task will run
in the service queue with a single processor.
The task executes with success.
- Only xml is generated.

Refs #832
Refs #2698

DavidHuber-NOAA committed:
Check that a PR driver is still running before trying to kill it (#2799)

Adds a check to the SSH command used to kill child PIDs of a defunct
driver instance on a different head node to prevent invalid kill
commands, preventing CI failures.

Resolves #2798

CoryMartin-NOAA committed:
Revert "remove CDUMP"

This reverts commit f82dc68f585cda20a2ecdbf9cf0841ed57209e94.

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
update gdas hash

CoryMartin-NOAA commented:
I don't understand what is going on with the conflict github is complaining about here...

DavidHuber-NOAA commented:
@CoryMartin-NOAA This is because the `GDASApp` hash you updated to is older than what is currently in develop.  Do you need this older hash?  If not, then from a console:
```console
> git clone git@github.com:CoryMartin-NOAA/global-workflow -b feature/snow-ensemble
> cd global-workflow
> git submodule update --init
> cd sorc/gdas.cd
> git checkout 67e8c5071d8b
> cd ..
> git add gdas.cd
> git commit -m "Revert GDASApp hash"
> git push
```

CoryMartin-NOAA commented:
hmm, somehow I must've screwed something up because I thought I committed the current develop hash. Let me dig in a bit

CoryMartin-NOAA committed:
change hash

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

AnilKumar-NOAA committed:
Add Gaea C5 to CI (#2814)

CI Testing and Jenkinsfile for Gaea C5

TerrenceMcGuinness-NOAA committed:
Jenkins Pipeline Updates (#2815)

Pipeline updates:
- get `gh` location on remote machine and define global bash env
variable `$GH` for GitHub CLI
- Failed cases are now displayed accordingly in the Jenkins dashboard
(see NOTE below)
-  Added the Build # in messaging for clarity when running from re-runs.
- Replaced Matrix construct for concurrency with the parallel method
that can use dynamic case lists
- With removing of the hard coded list of cases we now get list of cases
dynamically from the PR case directory
- See new look of dashboard below (has more annotations and displays
only used cases)

NOTE: **failFast** (quitting all cases on failing of one) still does not
work because it isn't quitting the running remote shells. We can make
this a configurable capability in feature request with some custom code.
The current behavior has the remaining cases continuing to run after a
FAIL label has been issued and it is incumbent of the code manager to
kill the CI job in the controller before resetting another Ready label.

---------

Co-authored-by: David Huber <69919478+DavidHuber-NOAA@users.noreply.github.com>

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA commented:
Please let me know what we still need to do to get this PR wrapped up

DavidHuber-NOAA commented on ush/python/pygfs/task/snowens_analysis.py:
```diff
@@ -0,0 +1,430 @@
+
+from wxflow import (AttrDict,
+                    FileHandler,
+                    to_fv3time, to_YMD, to_timedelta, add_to_datetime,
```
Nitpick: it doesn't look like `to_YMD` is used.

CoryMartin-NOAA replied:
updated thanks

DavidHuber-NOAA reviewed (dismissed):
Changes look good.

CoryMartin-NOAA committed:
remove to_ymdh

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA commented:
Hmm, need to figure out why it is using an absolute path here, because it is too long of a string for the code...

TerrenceMcGuinness-NOAA commented:
@CoryMartin-NOAA That path listed here in the PR message was created by the Jenkins Pipeline script and is not used in a "_string for the code_" I printed the absolute path here so you could find where the experiment directory is for the failed case.

BTW, If you have a CAC in a GFE you might be able see the Dash Board in Jenkins by following the links below labeled "Details".  You will be able to see that the other cases are still running.  If any of the others fail you will get another message like the one above.

WalterKolczynski-NOAA commented:
The code is looking for `fmsmpp` instead of `fmsmpp.nml`:
```
[38;21m2024-08-13 19:41:14,584 - INFO     - file_utils  : Copied /scratch1/NCEPDEV/global/CI/2690/gfs/fix/gdas/fv3jedi/fv3files/fmsmpp.nml to /scratch1/NCEPDEV/stmp2/Terry.McGuinness/RUNDIRS/C96C48_hybatmaerosnowDA_06d0fc25/gdas.2021122018/gdassnowanl_18/fv3jedi/fmsmpp.nml

...

0: FATAL from PE     0: get_ascii_file_num_lines: File /scratch1/NCEPDEV/stmp2/Terry.McGuinness/RUNDIRS/C96C48_hybatmaerosnowDA_06d0fc25/gdas.2021122018/gdassnowanl_18/fv3jedi/fmsmpp. does not exist.
```

CoryMartin-NOAA commented:
Yes thanks @WalterKolczynski-NOAA I'll need to change some YAMLs to relative paths instead of using $DATA, this will take me some time as it requires a GDAS submodule update. I'll ping when ready to re-test the CI.

WalterKolczynski-NOAA commented:
> Yes thanks @WalterKolczynski-NOAA I'll need to change some YAMLs to relative paths instead of using $DATA, this will take me some time as it requires a GDAS submodule update. I'll ping when ready to re-test the CI.

Not sure what relative vs absolute paths have to do with just an incorrect filename. The directory path was correct, just missing the `nml` at the end of the filename.

CoryMartin-NOAA commented:
it's exceeding the maximum character length in the FMS Fortran code. this is a known limitation of FMS with a maximum of 128 characters

WalterKolczynski-NOAA commented:
> it's exceeding the maximum character length in the FMS Fortran code. this is a known limitation of FMS with a maximum of 128 characters

🤦‍♂️

Looks like that is being expanded in 2024.3. Don't know why such an archaic length was in there to begin with.

CoryMartin-NOAA commented:
Changes in https://github.com/NOAA-EMC/GDASApp/pull/1250

CoryMartin-NOAA committed:
update gdas hash

CoryMartin-NOAA commented:
@WalterKolczynski-NOAA ok let's try this again, hash has been updated

TerrenceMcGuinness-NOAA commented:
```
Machine: Wcoss2
Start: Wed Aug 14 22:47:52 UTC 2024 on clogin03
---------------------------------------------------
Build: Completed at 08/14/24 11:26:33 PM
Case setup: Completed for experiment C48_ATM_6f732a19
Case setup: Skipped for experiment C48mx500_3DVarAOWCDA_6f732a19
Case setup: Skipped for experiment C48_S2SWA_gefs_6f732a19
Case setup: Completed for experiment C48_S2SW_6f732a19
Case setup: Completed for experiment C96_atm3DVar_extended_6f732a19
Case setup: Skipped for experiment C96_atm3DVar_6f732a19
Case setup: Completed for experiment C96C48_hybatmaerosnowDA_6f732a19
Case setup: Completed for experiment C96C48_hybatmDA_6f732a19
Case setup: Completed for experiment C96C48_ufs_hybatmDA_6f732a19
```

TerrenceMcGuinness-NOAA commented:
Main CI driver had a unrecoverable midstream failure on WCOSS2 due to connection failure to GitHub. 
Restarted next Running phase of CI workflow by hand.

```
                                                                 Req'd  Req'd   Elap
Job ID               Username Queue    Jobname    SessID NDS TSK Memory Time  S Time
-------------------- -------- -------- ---------- ------ --- --- ------ ----- - -----
148699244.cbqs01     terry.m* dev      C48_ATM_6*    --    1 128    --  03:00 R   -- 
148699245.cbqs01     terry.m* dev      C48_S2SW_*    --    1 128    --  03:00 R   -- 
```

CoryMartin-NOAA commented:
Looks like my wcoss2 APRUN is wrong... I'll take a look this afternoon to figure out what to change, unless anyone has any suggestions.

aerorahul commented:
> Looks like my wcoss2 APRUN is wrong... I'll take a look this afternoon to figure out what to change, unless anyone has any suggestions.

It seems you have used `srun` arguments for `mpiexec`
```bash
export APRUN_ESNOWRECEN="${APRUN} --cpus-per-task=${NTHREADS_ESNOWRECEN}"
```
should probably be
```bash
export APRUN_ESNOWRECEN="${APRUN} --ppn ${tasks_per_node} --cpu-bind-depth --depth=${NTHREADS_ESNOWRECEN}"
```
Please check that `tasks_per_node` and `NTHREADS_ESNOWRECEN` are correctly defined and of the right value.

CoryMartin-NOAA committed:
not worry about threads on wcoss

CoryMartin-NOAA commented:
Thanks @aerorahul I've just removed the threading, OOPS doesn't really use it anyways (yet)

RussTreadon-NOAA commented:
The `prepatmiodaobs` failures are due to a [bug](https://github.com/NOAA-EMC/GDASApp/pull/1246/files#r1717464548) in `bufr2ioda_apdupda.py`.  
```
Traceback (most recent call last):
  File "/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2690/global-workflow/sorc/gdas.cd/ush/ioda/bufr2ioda/bufr2ioda_adpupa.py", line 479, in <module>
    bufr_to_ioda(config, logger)
  File "/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2690/global-workflow/sorc/gdas.cd/ush/ioda/bufr2ioda/bufr2ioda_adpupa.py", line 209, in bufr_to_ioda
    specificHumidityOE = np.float32(np.ma.masked_array(np.full((len(qob)), 0.0)))
UnboundLocalError: local variable 'qob' referenced before assignment
```
A [fix](https://github.com/NOAA-EMC/GDASApp/pull/1252/commits/a3de92000a3fe239a5a4bdad51caa7b77e6285e5) has been added to GDASApp PR [#1252 ](https://github.com/NOAA-EMC/GDASApp/pull/1252) and pulled into g-w PR #2833.

Attention @PraveenKumar-NOAA , @CoryMartin-NOAA

CoryMartin-NOAA committed:
update gdas hash

CoryMartin-NOAA committed:
Merge branch 'develop' into feature/snow-ensemble

CoryMartin-NOAA commented:
The updated GDAS hash should hopefully fix the CI failure here

WalterKolczynski-NOAA commented:
@RussTreadon-NOAA can you take a look at why this and #2738 are both failing on the UFSDA atm test? If it doesn't have a quick fix, we may need to temporarily disable it to get v17 stuff in.

RussTreadon-NOAA commented:
`CoryMartin-NOAA:feature/snow-ensemble` points at [gdas.cd @ 0431b26](https://github.com/NOAA-EMC/GDASApp/tree/0431b26650c5e5d4eb741304a05c841d3fda0ddc).  This is the same GDASApp hash as PR #2833.  This GDASApp hash adds radiances to the atmospheric variational and local ensemble DA analysis jobs.  

g-w PR #2833 contains g-w changes needed to copy JEDI radiance bias correction files to COMROOT.  This PR does not contain these g-w changes.  As a result each of the init jobs failed because they can not copy the atms_n20 bias correction files
```
enkfgdasatmensanlinit.log:    raise OSError(f"unable to copy {source} to {target}")
enkfgdasatmensanlinit.log:OSError: unable to copy /lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2690/RUNTESTS/COMROOT/C96C48_ufs_hybatmDA_b3f8e5b3/gdas.20240223/18//analysis/atmos/gdas.t18z.atms_n20.satbias.nc to /lfs/h2/emc/stmp/terry.mcguinness/RUNDIRS/C96C48_ufs_hybatmDA_b3f8e5b3/enkfgdas.2024022400/enkfgdasatmensanl_00/obs/gdas.t18z.atms_n20.satbias.nc
gdasatmanlinit.log:    raise OSError(f"unable to copy {source} to {target}")
gdasatmanlinit.log:OSError: unable to copy /lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2690/RUNTESTS/COMROOT/C96C48_ufs_hybatmDA_b3f8e5b3/gdas.20240223/18//analysis/atmos/gdas.t18z.atms_n20.satbias.nc to /lfs/h2/emc/stmp/terry.mcguinness/RUNDIRS/C96C48_ufs_hybatmDA_b3f8e5b3/gdas.2024022400/gdasatmanl_00/obs/gdas.t18z.atms_n20.satbias.nc
gfsatmanlinit.log:    raise OSError(f"unable to copy {source} to {target}")
gfsatmanlinit.log:OSError: unable to copy /lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2690/RUNTESTS/COMROOT/C96C48_ufs_hybatmDA_b3f8e5b3/gdas.20240223/18//analysis/atmos/gdas.t18z.atms_n20.satbias.nc to /lfs/h2/emc/stmp/terry.mcguinness/RUNDIRS/C96C48_ufs_hybatmDA_b3f8e5b3/gfs.2024022400/gfsatmanl_00/obs/gdas.t18z.atms_n20.satbias.nc
```

We can not run g-w CI C96C48_ufs_hybatmDA until g-w PR #2833 is merged into g-w `develop`.  The other option is to back up the gdas.cd hash in this PR to a GDASApp hash before 0431b26.
```
0431b26 add lgetkf observer and solver templates (#1252)
925ebc8 update da-utils hash (#1255)
a58e705 Add DA-Utils to GDASApp (#1253)
eec4559 Add an entry to assimilate new snow dump files (sfcsno) replacing adpsfc_snow (#1245)
c5794ab fix for FMS paths that are too long in snow YAMLs (#1250)
9e38433 resolve inconsistencies between bufr2ioda python converters and jcb templates (#1241)
c68ef9d output gaussian increment for aerosols (#1248)
67e8c50 Add specificHumidity to ADPUPA BUFR DUMP Python API (#1246)
f3fa26d Stage atmospheric backgrounds and UFS cubed-sphere history files (#1236)
cc4568d Use ensemble member for mask in snow recentering (#1242)
```

WalterKolczynski-NOAA commented:
While #2833 is close, this one is a bit further along, so let's do this:

@CoryMartin-NOAA Please add `wcoss2` to the `skip_ci_on_hosts` list in [ci/cases/pr/C96C48_ufs_hybatmDA.yaml](https://github.com/NOAA-EMC/global-workflow/blob/develop/ci/cases/pr/C96C48_ufs_hybatmDA.yaml#L20)

@RussTreadon-NOAA Please turn the test back on as part of your PR after this one gets merged

CoryMartin-NOAA committed:
skip wcoss2 for ci for now

CoryMartin-NOAA commented:
@WalterKolczynski-NOAA done, thanks!

RussTreadon-NOAA commented:
@WalterKolczynski-NOAA , I'll try my best to remember to undo [1cb436e](https://github.com/NOAA-EMC/global-workflow/pull/2690/commits/1cb436e77b925ec51d15e279d788a671dba332df) in PR #2833 but I may forget.  Let's try to help each other to remember this important detail.
